Brokerages/Bitfinex/BitfinexBrokerage.Util.cs

```diff
@@ -0,0 +1,164 @@
+        {
+            public string BitfinexOrderType { get; set; }
+            public string Wallet { get; set; }
+            public OrderType OrderType { get; set; }
+        }
+
+        enum WalletType
+        {
+            exchange,
+            trading
+        }
+
+        //todo: trailing stop support
+        private static List<OrderTypeMap> _orderTypeMap = new List<OrderTypeMap>
+        {
+            { new OrderTypeMap { BitfinexOrderType = "exchange market", Wallet = WalletType.exchange.ToString(), OrderType = OrderType.Market } },
+            { new OrderTypeMap { BitfinexOrderType = "exchange limit", Wallet = WalletType.exchange.ToString(), OrderType = OrderType.Limit } },
+            { new OrderTypeMap { BitfinexOrderType = "exchange stop", Wallet = WalletType.exchange.ToString(), OrderType = OrderType.StopMarket } },
+            //{ new OrderTypeMap { BitfinexOrderType = "exchange trailing stop", Wallet = WalletType.exchange.ToString(), OrderType = OrderType.StopLimit } },
+
+            { new OrderTypeMap { BitfinexOrderType = "market", Wallet = WalletType.trading.ToString(), OrderType = OrderType.Market } },
+            { new OrderTypeMap { BitfinexOrderType = "limit", Wallet = WalletType.trading.ToString(), OrderType = OrderType.Limit } },
+            { new OrderTypeMap { BitfinexOrderType = "stop", Wallet = WalletType.trading.ToString(), OrderType = OrderType.StopMarket } },
+            //{ new OrderTypeMap { BitfinexOrderType = "trailing stop", Wallet = WalletType.trading.ToString(), OrderType = OrderType.StopLimit } },
+        };
+        #endregion
+
+
+        /// <summary>
+        /// Map exchange status
+        /// </summary>
+        /// <param name="response"></param>
+        /// <returns></returns>
+        public static OrderStatus MapOrderStatus(BitfinexOrderStatusResponse response)
+        {
+            decimal remainingAmount;
+            decimal executedAmount;
+            if (response.IsCancelled)
+            {
+                return OrderStatus.Canceled;
+            }
+            else if (decimal.TryParse(response.RemainingAmount, out remainingAmount) && remainingAmount > 0
+                && decimal.TryParse(response.ExecutedAmount, out executedAmount) && executedAmount > 0)
+            {
+                return OrderStatus.PartiallyFilled;
+            }
+            else if (response.IsLive)
+            {
+                return OrderStatus.Submitted;
+            }
+
+            return OrderStatus.Invalid;
+        }
+
+
+        /// <summary>
+        /// Map exchange order type
+        /// </summary>
+        /// <param name="orderType"></param>
+        /// <returns></returns>
+        public string MapOrderType(OrderType orderType)
+        {
+            var result = _orderTypeMap.Where(o => o.Wallet == Wallet && o.OrderType == orderType);
+
+            if (result != null & result.Count() == 1)
+            {
+                return result.Single().BitfinexOrderType;
```

Comment:
This is just for extra information, no change required as the impact here is extremely minimal., but typically you don't want to enumerate an `IEnumerable` more than once. The call to `result.Count()` will enumerate the entire enumerable and then `result.Single()` will re-enumerate it, albeit at most the first two entries. As I said, in this case it's not going to impact much , if anything at all. Here's an [example](http://twistedoakstudios.com/blog/Post7694_achieving-exponential-slowdown-by-enumerating-twice) of where it could really matter!